components/tidb_query_datatype/src/codec/mysql/vector.rs

```diff
@@ -89,33 +93,59 @@ pub trait VectorFloat32Decoder: NumberDecoder {
 impl<T: BufferReader> VectorFloat32Decoder for T {}
 
 /// Represents a reference of VectorFloat32 value aiming to reduce memory copy.
-#[derive(Clone, Copy, Eq, PartialEq, Ord, PartialOrd, Hash)]
+#[derive(Clone, Copy, Eq, PartialEq, Hash)]
 pub struct VectorFloat32Ref<'a> {
-    // Referred value
-    value: &'a [NotNan<f32>],
+    // Use &[u8] instead of &[u32] to allow reading from unaligned bytes. For example,
+    // bytes read from protobuf is usually not aligned by f32.
+    value: &'a [u8],
+}
+
+impl<'a> Ord for VectorFloat32Ref<'a> {
+    fn cmp(&self, other: &Self) -> Ordering {
+        let la = self.len();
+        let lb = other.len();
+        let common_len = std::cmp::min(la, lb);
+        for i in 0..common_len {
+            if self.index(i) > other.index(i) {
+                return Ordering::Greater;
+            } else if self.index(i) < other.index(i) {
+                return Ordering::Less;
+            }
+        }
+        la.cmp(&lb)
+    }
+}
+
+impl<'a> PartialOrd for VectorFloat32Ref<'a> {
+    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+        Some(self.cmp(other))
+    }
 }
 
 impl<'a> VectorFloat32Ref<'a> {
-    pub fn new(value: &[f32]) -> Result<VectorFloat32Ref<'_>> {
-        // Check again that NaN and ±Inf is not contained
-        for v in value {
-            if v.is_nan() {
+    pub fn new(value: &[u8]) -> Result<VectorFloat32Ref<'_>> {
+        if value.len() % 4 != 0 {
```

Comment:
How about changing all `4` to a const global variable? 
e.g. const F32_SIZE: usize = std::mem::size_of::<f32>();